File: src/swe/progress.rs
//! Background progress monitor for the SWE mining pipeline.
//!
//! Periodically logs pipeline statistics (candidates filtered, tasks extracted,
//! quality scored, tasks accepted) so operators can track long-running mining
//! runs without parsing individual log lines.

use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};

use tokio::task::JoinHandle;

/// Snapshot of pipeline progress counters at a point in time.
#[derive(Debug, Clone)]
pub struct ProgressSnapshot {
    /// Number of candidates that passed through the filter stage.
    pub filtered: usize,
    /// Number of tasks that passed extraction + test generation.
    pub extracted: usize,
    /// Number of tasks that were quality-scored.
    pub scored: usize,
    /// Number of tasks accepted into the final output.
    pub accepted: usize,
    /// Number of PRs enriched via GitHub API.
    pub enriched: usize,
    /// Number of PRs that completed pre-classification.
    pub preclassified: usize,
    /// Number of tasks currently in deep processing.
    pub deep_processing: usize,
    /// Number of active Docker containers.
    pub docker_active: usize,
    /// Wall-clock elapsed time since the monitor started.
    pub elapsed: Duration,
}

/// Shared atomic counters for pipeline progress tracking.
///
/// Cloned into pipeline worker tasks and incremented via `fetch_add`.
/// The background monitor reads these periodically to emit progress logs.
#[derive(Debug, Clone)]
pub struct ProgressCounters {
    /// Candidates evaluated by the filter stage.
    pub filtered: Arc<AtomicUsize>,
    /// Tasks that completed extraction + test generation.
    pub extracted: Arc<AtomicUsize>,
    /// Tasks that were quality-scored.
    pub scored: Arc<AtomicUsize>,
    /// Tasks accepted into the final output.
    pub accepted: Arc<AtomicUsize>,
    /// PRs enriched via GitHub API.
    pub enriched: Arc<AtomicUsize>,
    /// PRs that completed pre-classification.
    pub preclassified: Arc<AtomicUsize>,
    /// Tasks currently in deep processing.
    pub deep_processing: Arc<AtomicUsize>,
    /// Active Docker containers.
    pub docker_active: Arc<AtomicUsize>,
}

impl Default for ProgressCounters {
    fn default() -> Self {
        Self::new()
    }
}

impl ProgressCounters {
    /// Create a new set of zeroed progress counters.
    pub fn new() -> Self {
        Self {
            filtered: Arc::new(AtomicUsize::new(0)),
            extracted: Arc::new(AtomicUsize::new(0)),
            scored: Arc::new(AtomicUsize::new(0)),
            accepted: Arc::new(AtomicUsize::new(0)),
            enriched: Arc::new(AtomicUsize::new(0)),
            preclassified: Arc::new(AtomicUsize::new(0)),
            deep_processing: Arc::new(AtomicUsize::new(0)),
            docker_active: Arc::new(AtomicUsize::new(0)),
        }
    }

    /// Take a snapshot of the current counter values.
    pub fn snapshot(&self, start: Instant) -> ProgressSnapshot {
        ProgressSnapshot {
            filtered: self.filtered.load(Ordering::Relaxed),
            extracted: self.extracted.load(Ordering::Relaxed),
            scored: self.scored.load(Ordering::Relaxed),
            accepted: self.accepted.load(Ordering::Relaxed),
            enriched: self.enriched.load(Ordering::Relaxed),
            preclassified: self.preclassified.load(Ordering::Relaxed),
            deep_processing: self.deep_processing.load(Ordering::Relaxed),
            docker_active: self.docker_active.load(Ordering::Relaxed),
            elapsed: start.elapsed(),
        }
    }
}

/// A background task that periodically logs pipeline progress.
///
/// Spawns a tokio task that wakes every `interval` and logs a summary
/// of the pipeline counters. Call [`ProgressMonitor::stop`] to cancel.
pub struct ProgressMonitor {
    stop_flag: Arc<AtomicBool>,
    handle: Option<JoinHandle<()>>,
}

impl ProgressMonitor {
    /// Start a background progress monitor that logs every `interval`.
    ///
    /// # Arguments
    ///
    /// * `counters` - Shared atomic counters incremented by pipeline workers
    /// * `max_tasks` - Target number of tasks (used for progress percentage)
    /// * `interval` - How often to emit progress logs
    pub fn start(counters: ProgressCounters, max_tasks: usize, interval: Duration) -> Self {
        let stop_flag = Arc::new(AtomicBool::new(false));
        let flag = stop_flag.clone();
        let start = Instant::now();

        let handle = tokio::spawn(async move {
            let mut tick = tokio::time::interval(interval);
            tick.tick().await; // skip the immediate first tick

            loop {
                tick.tick().await;
                if flag.load(Ordering::Relaxed) {
                    break;
                }

                let snap = counters.snapshot(start);
                let pct = if max_tasks > 0 {
                    (snap.accepted as f64 / max_tasks as f64 * 100.0).min(100.0)
                } else {
                    0.0
                };

                let elapsed_secs = snap.elapsed.as_secs_f64();
                let enriched_per_sec = if elapsed_secs > 0.0 {
                    snap.enriched as f64 / elapsed_secs
                } else {
                    0.0
                };
                let filtered_per_sec = if elapsed_secs > 0.0 {
                    snap.filtered as f64 / elapsed_secs
                } else {
                    0.0
                };
                let accepted_per_sec = if elapsed_secs > 0.0 {
                    snap.accepted as f64 / elapsed_secs
                } else {
                    0.0
                };

                tracing::info!(
                    filtered = snap.filtered,
                    extracted = snap.extracted,
                    scored = snap.scored,
                    accepted = snap.accepted,
                    enriched = snap.enriched,
                    preclassified = snap.preclassified,
                    deep_processing = snap.deep_processing,
                    docker_active = snap.docker_active,
                    max_tasks = max_tasks,
                    progress_pct = format!("{:.1}%", pct),
                    elapsed_secs = snap.elapsed.as_secs(),
                    enriched_per_sec = format!("{:.2}", enriched_per_sec),
                    filtered_per_sec = format!("{:.2}", filtered_per_sec),
                    accepted_per_sec = format!("{:.4}", accepted_per_sec),
                    "Pipeline progress"
                );
            }
        });

        Self {
            stop_flag,
            handle: Some(handle),
        }
    }

    /// Signal the background monitor to stop and wait for it to finish.
    pub async fn stop(mut self) {
        self.stop_flag.store(true, Ordering::Relaxed);
        if let Some(handle) = self.handle.take() {
            let _ = handle.await;
        }
    }
}

impl Drop for ProgressMonitor {
    fn drop(&mut self) {
        self.stop_flag.store(true, Ordering::Relaxed);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_progress_counters_default() {
        let counters = ProgressCounters::new();
        let snap = counters.snapshot(Instant::now());
        assert_eq!(snap.filtered, 0);
        assert_eq!(snap.extracted, 0);
        assert_eq!(snap.scored, 0);
        assert_eq!(snap.accepted, 0);
        assert_eq!(snap.enriched, 0);
        assert_eq!(snap.preclassified, 0);
        assert_eq!(snap.deep_processing, 0);
        assert_eq!(snap.docker_active, 0);
    }

    #[test]
    fn test_progress_counters_increment() {
        let counters = ProgressCounters::new();
        counters.filtered.fetch_add(10, Ordering::Relaxed);
        counters.extracted.fetch_add(5, Ordering::Relaxed);
        counters.scored.fetch_add(3, Ordering::Relaxed);
        counters.accepted.fetch_add(1, Ordering::Relaxed);
        counters.enriched.fetch_add(20, Ordering::Relaxed);
        counters.preclassified.fetch_add(15, Ordering::Relaxed);
        counters.deep_processing.fetch_add(4, Ordering::Relaxed);
        counters.docker_active.fetch_add(2, Ordering::Relaxed);

        let snap = counters.snapshot(Instant::now());
        assert_eq!(snap.filtered, 10);
        assert_eq!(snap.extracted, 5);
        assert_eq!(snap.scored, 3);
        assert_eq!(snap.accepted, 1);
        assert_eq!(snap.enriched, 20);
        assert_eq!(snap.preclassified, 15);
        assert_eq!(snap.deep_processing, 4);
        assert_eq!(snap.docker_active, 2);
    }

    #[test]
    fn test_progress_counters_clone_shares_state() {
        let counters = ProgressCounters::new();
        let clone = counters.clone();

        counters.accepted.fetch_add(1, Ordering::Relaxed);
        assert_eq!(clone.accepted.load(Ordering::Relaxed), 1);
    }

    #[tokio::test]
    async fn test_progress_monitor_start_stop() {
        let counters = ProgressCounters::new();
        counters.accepted.fetch_add(3, Ordering::Relaxed);

        let monitor = ProgressMonitor::start(counters, 10, Duration::from_millis(50));

        tokio::time::sleep(Duration::from_millis(120)).await;
        monitor.stop().await;
    }
}

File: src/swe_forge/swe/progress.py
"""Progress monitoring for SWE task processing pipeline.

Thread-safe counters and background monitor for tracking processing progress.
"""

from __future__ import annotations

import asyncio
import threading
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

import structlog

if TYPE_CHECKING:
    pass

logger = structlog.get_logger(__name__)


@dataclass
class ProgressSnapshot:
    """Immutable snapshot of progress counters at a point in time."""

    filtered: int = 0
    extracted: int = 0
    scored: int = 0
    accepted: int = 0
    enriched: int = 0
    preclassified: int = 0
    deep_processing: int = 0
    docker_active: int = 0
    elapsed_seconds: float = 0.0


class ProgressCounters:
    """Thread-safe progress counters using threading.Lock.

    Tracks various processing stages for SWE task pipeline.
    Counter increment timing:
    - `enriched`: After successful GitHub API enrichment
    - `filtered`: After filter stage (passed + rejected)
    - `preclassified`: After pre-classification (cached or fresh)
    - `scored`: After quality assessment
    - `accepted`: After task accepted into final output
    - `extracted`: When added to tasks list
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._counters: dict[str, int] = {
            "filtered": 0,
            "extracted": 0,
            "scored": 0,
            "accepted": 0,
            "enriched": 0,
            "preclassified": 0,
            "deep_processing": 0,
            "docker_active": 0,
        }

    def increment(self, name: str, delta: int = 1) -> None:
        """Increment a counter by the given delta.

        Args:
            name: Counter name (filtered, extracted, scored, accepted, enriched,
                  preclassified, deep_processing, docker_active).
            delta: Amount to increment by (default: 1).

        Raises:
            KeyError: If counter name is invalid.
        """
        with self._lock:
            if name not in self._counters:
                raise KeyError(f"Unknown counter: {name}")
            self._counters[name] += delta

    def get(self, name: str) -> int:
        """Get current value of a counter.

        Args:
            name: Counter name.

        Returns:
            Current counter value.

        Raises:
            KeyError: If counter name is invalid.
        """
        with self._lock:
            if name not in self._counters:
                raise KeyError(f"Unknown counter: {name}")
            return self._counters[name]

    def snapshot(self, start_time: float) -> ProgressSnapshot:
        """Create an immutable snapshot of all counters.

        Args:
            start_time: Start time as returned by time.time().

        Returns:
            ProgressSnapshot with current counter values and elapsed time.
        """
        elapsed = time.time() - start_time
        with self._lock:
            return ProgressSnapshot(
                filtered=self._counters["filtered"],
                extracted=self._counters["extracted"],
                scored=self._counters["scored"],
                accepted=self._counters["accepted"],
                enriched=self._counters["enriched"],
                preclassified=self._counters["preclassified"],
                deep_processing=self._counters["deep_processing"],
                docker_active=self._counters["docker_active"],
                elapsed_seconds=elapsed,
            )


class ProgressMonitor:
    """Background task that logs progress at regular intervals.

    Logs structured progress message every interval seconds with:
    - Counter values
    - Elapsed time
    - Tasks per second throughput
    """

    def __init__(self) -> None:
        self._task: asyncio.Task[None] | None = None
        self._stop_event = asyncio.Event()

    @classmethod
    def start(
        cls,
        counters: ProgressCounters,
        max_tasks: int,
        interval: float = 30.0,
    ) -> ProgressMonitor:
        """Create and start a progress monitor.

        Args:
            counters: ProgressCounters instance to monitor.
            max_tasks: Expected total number of tasks.
            interval: Logging interval in seconds (default: 30.0).

        Returns:
            Running ProgressMonitor instance.
        """
        monitor = cls()
        start_time = time.time()
        monitor._task = asyncio.create_task(
            monitor._monitor_loop(counters, start_time, max_tasks, interval)
        )
        return monitor

    async def _monitor_loop(
        self,
        counters: ProgressCounters,
        start_time: float,
        max_tasks: int,
        interval: float,
    ) -> None:
        """Monitor loop that logs progress periodically.

        Args:
            counters: ProgressCounters to monitor.
            start_time: Start time for elapsed calculation.
            max_tasks: Expected total number of tasks.
            interval: Logging interval in seconds.
        """
        while not self._stop_event.is_set():
            try:
                await asyncio.wait_for(self._stop_event.wait(), timeout=interval)
                # stop_event was set, exit loop
                break
            except asyncio.TimeoutError:
                # Timeout reached, emit progress
                snap = counters.snapshot(start_time)
                self._emit_progress(snap, max_tasks)

    def _emit_progress(self, snap: ProgressSnapshot, max_tasks: int) -> None:
        """Emit structured progress log message.

        Args:
            snap: ProgressSnapshot with current values.
            max_tasks: Expected total number of tasks.
        """
        tasks_per_sec = (
            snap.filtered / snap.elapsed_seconds if snap.elapsed_seconds > 0 else 0.0
        )

        logger.info(
            "pipeline_progress",
            filtered=snap.filtered,
            extracted=snap.extracted,
            scored=snap.scored,
            accepted=snap.accepted,
            enriched=snap.enriched,
            preclassified=snap.preclassified,
            deep_processing=snap.deep_processing,
            docker_active=snap.docker_active,
            max_tasks=max_tasks,
            elapsed_seconds=round(snap.elapsed_seconds, 2),
            tasks_per_second=round(tasks_per_sec, 2),
        )

    async def stop(self) -> None:
        """Stop the progress monitor and wait for it to finish."""
        self._stop_event.set()
        if self._task is not None:
            try:
                await self._task
            except asyncio.CancelledError:
                pass


__all__ = [
    "ProgressSnapshot",
    "ProgressCounters",
    "ProgressMonitor",
]
